[queues] YT-23208: Queue static table exports refactor

ItIsApachee opened the pull request:
Mirroring PR for reviewing from github

ItIsApachee commented:
Need to add small fix to how `accumulatedMinFragmentUnixTs` is calculated. Currently it would break everything if some data is flushed that corresponds to unix ts which has already been exported. I'll do it tomorrow.

achulkov2 reviewed (changes requested):
Left some comments here and there, generally I think the code became a bit too confusing and we should take some time to make this PR cleaner.

ItIsApachee commented on yt/yt/server/queue_agent/queue_static_table_exporter.cpp:
```diff
@@ -160,7 +189,7 @@ class TQueueExportTask
             .NodeId;
 
-        // Take it for guarantee that only one queue agent instance do this export.
+        // Take it for granted that only one queue agent instance do this export.
```
It is not possible in a sense that there can't be 2 concurrent transactions doing the same export since we take the same shared lock. It might be possible if leader switch and config change occurred at the same time, but I don't think that is a problem.
Do you agree with that? Or am I missing something?
If you do agree, then I'll re-write the comment to reflect this behavior.

ItIsApachee commented on yt/yt/server/queue_agent/config.cpp:
```diff
@@ -88,6 +92,14 @@ void TQueueControllerDynamicConfig::Register(TRegistrar registrar)
 
+void TQueueStaticTableExportManagerDynamicConfig::Register(TRegistrar registrar)
+{
+    registrar.Parameter("max_concurrent_exports", &TThis::MaxConcurrentExports)
```
I added TQueueStaticTableExportManager, but forgot to stage new files for github. Fixed it already as you can see from the latest commit.

ItIsApachee commented on yt/yt/server/queue_agent/queue_static_table_exporter.cpp:
```diff
@@ -241,45 +316,378 @@ class TQueueExportTask
+        const TTableSchemaPtr& QueueSchema;
+        const TMasterTableSchemaId QueueSchemaId;
+
+        const TQueueStaticExportConfig ExportConfig;
```
Well, I thought that some values are better of copied rather than stored as a pointer or a reference to the actual data, e.g. primitive types. To be completely honest, I'm not that sure how far this argument extends, and at what point do we prefer references or pointers rather than a copy.
In the case of `TQueueStaticExportConfig` I thought that maybe reference/pointer would be better, but at this point should this really be a lite yson struct? If we care enough to save some space for this config, shouldn't we make it ref-counted yson struct?

ItIsApachee replied:
Are you against making a copy of `TMasterTableSchemaId` (just a guid)? And did you meaningfully add this comment for the first line? Is there a problem with `const TTableSchemaPtr& QueueSchema;`?

savnadya commented on yt/yt/server/queue_agent/queue_static_table_export_manager.cpp:
```diff
@@ -0,0 +1,71 @@
+TQueueStaticTableExportManager::TQueueStaticTableExportManager(
+    TQueueStaticTableExportManagerDynamicConfig dynamicConfig,
+    const IInvokerPtr& underlyingInvoker)
+    : DynamicConfig_(dynamicConfig)
```
std::move

savnadya commented on yt/yt/server/queue_agent/queue_static_table_export_manager.cpp:
```diff
@@ -0,0 +1,71 @@
+    TQueueStaticTableExportManagerDynamicConfig dynamicConfig,
+    const IInvokerPtr& underlyingInvoker)
+    : DynamicConfig_(dynamicConfig)
+    , Invoker_(CreateBoundedConcurrencyInvoker(underlyingInvoker, DynamicConfig_.MaxConcurrentExports))
```
MaxConcurrentExportCount?

savnadya commented on yt/yt/server/queue_agent/config.h:
```diff
@@ -159,6 +180,9 @@ class TQueueAgentDynamicConfig
     TQueueControllerDynamicConfigPtr Controller;
 
+    //! Configuration of queue static table export manager.
+    TQueueStaticTableExportManagerDynamicConfig QueueStaticTableExportManager;
```
Let's store it as intrusive ptr?
`TQueueStaticTableExportManagerDynamicConfigPtr QueueStaticTableExportManager`

savnadya commented on yt/yt/server/queue_agent/queue_static_table_exporter.cpp:
```diff
@@ -211,17 +278,25 @@ class TQueueExportTask
+        return LastSubtaskError_;
+    }
+
+    TQueueExportProgressPtr GetExportProgress() const {
```
And here

savnadya commented on yt/yt/server/queue_agent/queue_static_table_exporter.cpp:
```diff
@@ -241,45 +316,378 @@ class TQueueExportTask
+        //! Data statistics collected from attaching chunks.
+        TDataStatistics DataStatistics;
+
+        TQueueExportSubtask(
```
`struct` in our code base usually don't have a constructor or public methods. It looks like a `class`.

savnadya commented on yt/yt/server/queue_agent/queue_static_table_exporter.cpp:
```diff
@@ -241,45 +316,378 @@ class TQueueExportTask
+            , Logger(logger)
+        { }
+
+        TError Run() {
```
Move `{` to the next line

savnadya commented on yt/yt/server/queue_agent/queue_static_table_exporter.cpp:
```diff
@@ -241,45 +316,378 @@ class TQueueExportTask
+            return TError();
+        }
+
+        void GuardedRun() {
```
And here

achulkov2 commented on yt/yt/server/queue_agent/queue_static_table_export_manager.cpp:
```diff
@@ -0,0 +1,76 @@
+    const TQueueStaticTableExportManagerDynamicConfigPtr& /*oldConfig*/,
+    const TQueueStaticTableExportManagerDynamicConfigPtr& newConfig)
+{
+    Throttler_->Reconfigure(TThroughputThrottlerConfig::Create(newConfig->ExportRateLimit));
```
Let's log the change.

achulkov2 commented on yt/yt/server/queue_agent/queue_static_table_exporter.cpp:
```diff
@@ -143,8 +593,9 @@ class TQueueExportTask
-    TQueueExportProgressPtr Run()
+    void DoRun()
     {
+        // XXX(apachee): Do we need this now?
```
I am not sure removing this will not break tests :) 
Interesting why I didn't write a comment though, I barely remember the reasoning.

achulkov2 commented on yt/yt/server/queue_agent/queue_static_table_exporter.cpp:
```diff
@@ -770,63 +1039,76 @@ TQueueExporter::TQueueExporter(
+
+    auto guard = Guard(Lock_);
+    auto exportUnixTs = GetExportFragmentUnixTs(now.Seconds(), ExportConfig_->ExportPeriod);
+    guard.Release();
```
Let's introduce some method for getting the whole config under lock, and use it here and in other places to avoid manually releasing guards. But you will need to deep-copy, since dynamic config changes can happen.

achulkov2 commented on yt/yt/server/queue_agent/queue_static_table_exporter.cpp:
```diff
@@ -838,24 +1120,88 @@ void TQueueExporter::Export()
+        }
     }
+
+    guard.Release();
```
My personal preference with guards is scoping. If you are not doing that, you might as well use plain `Lock_.Acquire()` and `Lock_.Release()`.
But I think we tend to prefer this:
```
{
    auto guard = Guard(Lock_);
    // Do something.
}
```

achulkov2 commented on yt/yt/server/queue_agent/queue_static_table_exporter.cpp:
```diff
@@ -838,24 +1120,88 @@ void TQueueExporter::Export()
-    auto config = GetConfig();
+    // XXX(apachee): Throttle inside export task?
+    auto throttler = QueueStaticTableExportManager_->GetExportThrottler();
+    WaitFor(throttler->Throttle(1))
```
I think we should account 1 per each exported table.

achulkov2 reviewed (changes requested):
First batch of comments.

ItIsApachee commented on yt/yt/server/queue_agent/queue_static_table_exporter.cpp:
```diff
@@ -750,18 +967,70 @@ TQueueExportProfilingCounters::TQueueExportProfilingCounters(const TProfiler& pr
+    void OnDynamicConfigChanged(const TQueueExporterDynamicConfig& newDynamicConfig) override;
+
+private:
+    const TString ExportName_;
+    const NQueueClient::TCrossClusterReference Queue_;
+
+    TSpinLock Lock_;
+    TQueueStaticExportConfigPtr ExportConfig_;
+    TQueueExporterDynamicConfig DynamicConfig_;
+    TQueueExportProgressPtr ExportProgress_;
+    //! Export unix timestamp corresponding to the last successful export.
+    ui64 LastSuccessfulExportUnixTs = 0;
+
+    const NHiveClient::TClientDirectoryPtr ClientDirectory_;
+    const IInvokerPtr Invoker_;
+    const IQueueStaticTableExportManagerPtr QueueStaticTableExportManager_;
+    const NAlertManager::IAlertCollectorPtr AlertCollector_;
+    const TQueueExportProfilingCountersPtr ProfilingCounters_;
+
+    const NConcurrency::TPeriodicExecutorPtr Executor_;
+
+    const NLogging::TLogger Logger;
```
Re-order fields such that const fields are grouped together

ItIsApachee commented on yt/yt/server/queue_agent/queue_static_table_exporter.cpp:
```diff
@@ -838,24 +1120,88 @@ void TQueueExporter::Export()
+            Invoker_,
+            ProfilingCounters_,
+            Queue_.Path,
+            ExportConfig_,
```
Do we really need to deep-copy this config? I think it is okay to assume that the only changes to config are those  which replace one pointer for another, without changing the value pointers point to.
Maybe this invariant should somehow be enforced? In some code we have `TIntrusivePtr<const TType>`, maybe that can be used here?

ItIsApachee commented on yt/yt/server/queue_agent/queue_static_table_exporter.cpp:
```diff
@@ -207,79 +681,43 @@ class TQueueExportTask
         ProfilingCounters_->ExportedChunks.Increment(diff.ChunkCount);
-        ProfilingCounters_->ExportedTables.Increment();
+        ProfilingCounters_->ExportedTables.Increment(completedSubtaskCount);
+        ProfilingCounters_->UncompletedExports.Increment(failedSubtaskCount);
```
What if we use "skipped" terminology here to clearly communicate what is happening?

achulkov2 commented on yt/yt/server/queue_agent/queue_static_table_exporter.h:
```diff
@@ -69,6 +75,8 @@ struct TQueueExportProfilingCounters final
     NProfiling::TCounter ExportedRows;
     NProfiling::TCounter ExportedChunks;
     NProfiling::TCounter ExportedTables;
+    NProfiling::TCounter SkippedTables;
```
Let's add comments for what these mean.

achulkov2 commented on yt/yt/tests/integration/queues/test_queue_agent.py:
```diff
@@ -3563,30 +3613,34 @@ def test_export_iteration_with_no_data(self):
             },
         })
 
+        # COMPAT(apachee): We check "last_successful_export_iteration_instant" until third parties rely on this field.
```
stop relying?

achulkov2 commented on yt/yt/tests/integration/queues/test_queue_agent.py:
```diff
@@ -3235,6 +3146,142 @@ def test_multicell_export(self, queue_external_cell_tag):
+        self._sleep_until_next_export_instant(export_period_seconds)
+        time.sleep(export_period_seconds)
+
+        # XXX(apachee): Is it really enough to check that retry is scheduled way before the next export iteration?
```
let's add a test where we wait longer, write/flush some more data and then see how it has to retry creating the first table and also create a new one

achulkov2 commented on yt/yt/tests/integration/queues/test_queue_agent.py:
```diff
@@ -3235,6 +3146,142 @@ def test_multicell_export(self, queue_external_cell_tag):
+        self._check_export(export_dir, [["42"] * 2, ["13"] * 2])
+
+        self.remove_export_destination(export_dir)
+
```
More test ideas:
* Check behaviour of partial export tasks, where only some exports are successful, both to some natural causes (idk how to do it yet, maybe set some quota limits somehow? like set a separate account and a low node quota for it) and configuration limits (max exported tables per task). And also check multiple iterations here, so that tables are not re-exported for example. Also check alerts are propagated correctly here.
* Check ttl depending on how you implement it.
* Check throttler? It is a bit annoying to do, but I recently did it for master r/w user throttlers and I think it is a good practice to write tests for throttlers where possible. 

First one I am strongly requesting.

achulkov2 commented on yt/yt/server/queue_agent/queue_static_table_exporter.cpp:
```diff
@@ -209,79 +363,124 @@ class TQueueExportTask
         ProfilingCounters_->ExportedChunks.Increment(diff.ChunkCount);
-        ProfilingCounters_->ExportedTables.Increment();
+        ProfilingCounters_->ExportedTables.Increment(exportedTableCount);
+        ProfilingCounters_->SkippedTables.Increment(skippedTableCount);
```
Can we add a gauge for number of "pending" exports? Say, if exports are failing for 5 export instants already, it should be equal to 5. 

~queue_size for a lot of our services/throttlers

savnadya commented on yt/yt/server/queue_agent/queue_agent.cpp:
```diff
@@ -7,6 +7,7 @@
 #include "object.h"
 #include "queue_controller.h"
 #include "queue_static_table_exporter.h"
+#include "queue_static_table_export_manager.h"
```
Wrong order

savnadya commented on yt/yt/server/queue_agent/queue_static_table_exporter.cpp:
```diff
@@ -764,112 +958,231 @@ TQueueExportProfilingCounters::TQueueExportProfilingCounters(const TProfiler& pr
-        : std::nullopt);
-}
+public:
+    TQueueStaticTableExporter(
```
Let's name it `TQueueExporter`?
`TQueueExporterDynamicConfig` and `TQueueExportTask` are without `StaticTable`

savnadya commented on yt/yt/server/queue_agent/queue_static_table_exporter.cpp:
```diff
@@ -764,112 +958,231 @@ TQueueExportProfilingCounters::TQueueExportProfilingCounters(const TProfiler& pr
+    ui64 LastSuccessfulExportUnixTs_ = 0;
+
+    const TString ExportName_;
+    const NQueueClient::TCrossClusterReference Queue_;
```
Namespace isn't needed here, and below

ItIsApachee commented:
Sadly, due to refactoring, mirroring diff to github was problematic, and file renaming was not properly recognized, even though I had this in git status:
```
        renamed:    yt/yt/server/queue_agent/queue_static_table_export_manager.h -> yt/yt/server/queue_agent/queue_export_manager.h
        renamed:    yt/yt/server/queue_agent/queue_static_table_exporter.cpp -> yt/yt/server/queue_agent/queue_exporter.cpp
        renamed:    yt/yt/server/queue_agent/queue_static_table_exporter.h -> yt/yt/server/queue_agent/queue_exporter.h
```

achulkov2 commented on yt/yt/server/queue_agent/queue_exporter.cpp:
```diff
@@ -0,0 +1,1221 @@
+
+        ui64 timeLagSeconds = 0;
+        auto exportProgress = GetExportProgress();
+        if (exportProgress->LastExportUnixTs <= exportUnixTsUpperBound) {
```
Is it possible for `exportProgress->LastExportUnixTs` to be larger than `exportUnixTsUpperBound`?

ItIsApachee replied:
I think it can be the case if there are some issues with system time, but, realistically, I just guarded against unsigned integer overflow to prevent weird values for metrics.

achulkov2 replied:
Maybe `YT_LOG_ALERT` for such cases?

achulkov2 commented on yt/yt/tests/library/yt_queue_agent_test_base.py:
```diff
@@ -91,6 +92,24 @@ class ObjectAlertHelper:
     def __init__(self, alerts):
         self.alerts = alerts
 
+    def check_matching(self, category, text=None, attributes=None):
```
I would prefer it if we shared code with the helper below. Either put assert vs raise with some callback, or maybe just return False in these cases and outward assert will fail?

achulkov2 reviewed (approved):
LGTM, left 2 small comments, you can just leave TODOs for them, if you don't want to fix.
